File: app/api/routes/villains.py
"""
反派管理 API 路由
GodView v9: 反派与冲突系统专用接口
"""

import logging
from datetime import datetime
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from app.services.skill_service import get_skill_service, ExecuteSkillDTO

logger = logging.getLogger(__name__)

router = APIRouter()


# ==================== 数据模型 ====================

class VillainLevel(str):
    """反派层级"""
    SMALL = "small"       # 小反派
    MEDIUM = "medium"     # 中BOSS
    MAJOR = "major"       # 大BOSS


class VillainStatus(str):
    """反派状态"""
    ACTIVE = "active"
    DEFEATED = "defeated"
    REDEEMED = "redeemed"
    ESCAPED = "escaped"


class ConflictType(str):
    """冲突类型"""
    MAIN = "main"         # 主线冲突
    SUB = "sub"           # 支线冲突
    HIDDEN = "hidden"     # 隐藏冲突


# ==================== 请求/响应模型 ====================

class CreateVillainRequest(BaseModel):
    """创建反派请求"""
    project_id: str
    name: str = Field(..., description="反派名称")
    alias: Optional[str] = Field(None, description="反派别称")
    villain_level: str = Field(default="small", description="反派层级：small/medium/major")
    description: Optional[str] = Field(None, description="反派描述")
    motivation: Optional[str] = Field(None, description="反派动机")
    power_level: Optional[str] = Field(None, description="实力等级")


class VillainResponse(BaseModel):
    """反派响应"""
    success: bool
    villain: Dict[str, Any]
    conflict_suggestions: List[str]


class UpdateVillainRequest(BaseModel):
    """更新反派请求"""
    name: Optional[str] = None
    alias: Optional[str] = None
    description: Optional[str] = None
    motivation: Optional[str] = None
    power_level: Optional[str] = None
    status: Optional[str] = None


class VillainDefeatRequest(BaseModel):
    """反派被击败请求"""
    chapter_number: int
    defeat_method: str = Field(..., description="击败方式")
    aftermath: Optional[str] = Field(None, description="后续影响")


class CreateConflictRequest(BaseModel):
    """创建冲突线请求"""
    project_id: str
    title: str = Field(..., description="冲突标题")
    conflict_type: str = Field(default="main", description="冲突类型：main/sub/hidden")
    description: Optional[str] = Field(None, description="冲突描述")
    parties: List[str] = Field(default_factory=list, description="参与方")
    villain_id: Optional[str] = Field(None, description="关联反派")


class ConflictResponse(BaseModel):
    """冲突线响应"""
    success: bool
    conflict: Dict[str, Any]
    escalation_plan: List[str]


class EscalateConflictRequest(BaseModel):
    """冲突升级请求"""
    chapter_number: int
    escalation_type: str = Field(..., description="升级类型")
    description: str = Field(..., description="升级描述")


class VillainDesignRequest(BaseModel):
    """反派设计请求"""
    project_id: str
    villain_level: str = Field(default="major", description="反派层级")
    genre: Optional[str] = Field(None, description="题材类型")
    protagonist_power: Optional[str] = Field(None, description="主角实力")
    story_phase: Optional[str] = Field(None, description="故事阶段")


class VillainDesignResponse(BaseModel):
    """反派设计响应"""
    success: bool
    villain_design: Dict[str, Any]
    conflict_design: Dict[str, Any]
    defeat_timeline: Dict[str, Any]
    suggestions: List[str]


class ConflictTrackingRequest(BaseModel):
    """冲突追踪请求"""
    project_id: str
    current_chapter: int
    active_conflicts: Optional[List[str]] = None


class ConflictTrackingResponse(BaseModel):
    """冲突追踪响应"""
    success: bool
    conflict_status: List[Dict[str, Any]]
    escalation_warnings: List[str]
    resolution_suggestions: List[str]


# ==================== 反派 API ====================

@router.post("/design", response_model=VillainDesignResponse)
async def design_villain(request: VillainDesignRequest):
    """
    设计反派角色

    根据题材和故事阶段，生成完整的反派设计方案
    """
    skill_service = get_skill_service()

    result = await skill_service.execute_skill(ExecuteSkillDTO(
        skill_id="skill_villain_management",
        project_id=request.project_id,
        parameters={
            "task": "design",
            "villain_level": request.villain_level,
            "genre": request.genre or "玄幻",
            "protagonist_power": request.protagonist_power or "",
            "story_phase": request.story_phase or "前期",
        }
    ))

    if not result.success:
        raise HTTPException(status_code=500, detail=result.error)

    import json
    try:
        data = json.loads(result.output)
        return VillainDesignResponse(
            success=True,
            villain_design=data.get("villain_design", {}),
            conflict_design=data.get("conflict_design", {}),
            defeat_timeline=data.get("defeat_timeline", {}),
            suggestions=data.get("suggestions", []),
        )
    except json.JSONDecodeError:
        raise HTTPException(status_code=500, detail="解析设计结果失败")


@router.post("", response_model=VillainResponse)
async def create_villain(request: CreateVillainRequest):
    """
    创建反派

    创建新的反派角色
    """
    # TODO: 实现数据库存储
    villain_id = f"villain_{datetime.now().strftime('%Y%m%d%H%M%S')}"

    villain = {
        "id": villain_id,
        "project_id": request.project_id,
        "name": request.name,
        "alias": request.alias,
        "villain_level": request.villain_level,
        "description": request.description or "",
        "motivation": request.motivation or "",
        "power_level": request.power_level or "",
        "status": "active",
        "created_at": datetime.now().isoformat(),
    }

    return VillainResponse(
        success=True,
        villain=villain,
        conflict_suggestions=[
            f"建议为{request.name}设计一个与主角的核心冲突",
            "考虑反派的动机如何与主角目标对立",
        ]
    )


@router.get("/{villain_id}")
async def get_villain(villain_id: str, project_id: str):
    """
    获取反派详情
    """
    # TODO: 从数据库获取
    raise HTTPException(status_code=404, detail="反派不存在")


@router.put("/{villain_id}")
async def update_villain(villain_id: str, request: UpdateVillainRequest):
    """
    更新反派信息
    """
    # TODO: 实现数据库更新
    return {"success": True, "message": f"反派 {villain_id} 已更新"}


@router.post("/{villain_id}/defeat")
async def defeat_villain(villain_id: str, request: VillainDefeatRequest):
    """
    标记反派被击败

    记录反派的退场方式和后续影响
    """
    # TODO: 实现数据库更新
    return {
        "success": True,
        "villain_id": villain_id,
        "defeat_chapter": request.chapter_number,
        "defeat_method": request.defeat_method,
        "status": "defeated",
    }


# ==================== 冲突线 API ====================

@router.post("/conflicts", response_model=ConflictResponse)
async def create_conflict(request: CreateConflictRequest):
    """
    创建冲突线

    创建新的故事冲突线
    """
    conflict_id = f"conflict_{datetime.now().strftime('%Y%m%d%H%M%S')}"

    conflict = {
        "id": conflict_id,
        "project_id": request.project_id,
        "title": request.title,
        "conflict_type": request.conflict_type,
        "description": request.description or "",
        "parties": request.parties,
        "villain_id": request.villain_id,
        "status": "active",
        "escalation_history": [],
        "created_at": datetime.now().isoformat(),
    }

    return ConflictResponse(
        success=True,
        conflict=conflict,
        escalation_plan=[
            "第一阶段：冲突萌芽",
            "第二阶段：矛盾激化",
            "第三阶段：正面冲突",
            "第四阶段：高潮对决",
        ]
    )


@router.get("/conflicts")
async def list_conflicts(project_id: str, conflict_type: Optional[str] = None):
    """
    获取冲突线列表

    列出项目的所有冲突线
    """
    # TODO: 从数据库获取
    return {
        "conflicts": [],
        "total": 0,
    }


@router.post("/conflicts/{conflict_id}/escalate")
async def escalate_conflict(conflict_id: str, request: EscalateConflictRequest):
    """
    升级冲突

    记录冲突的升级事件
    """
    return {
        "success": True,
        "conflict_id": conflict_id,
        "escalation": {
            "chapter": request.chapter_number,
            "type": request.escalation_type,
            "description": request.description,
            "escalated_at": datetime.now().isoformat(),
        }
    }


@router.post("/tracking", response_model=ConflictTrackingResponse)
async def track_conflicts(request: ConflictTrackingRequest):
    """
    冲突追踪

    分析当前冲突状态，提供升级警告和解决建议
    """
    skill_service = get_skill_service()

    result = await skill_service.execute_skill(ExecuteSkillDTO(
        skill_id="skill_villain_management",
        project_id=request.project_id,
        parameters={
            "task": "tracking",
            "current_chapter": request.current_chapter,
            "active_conflicts": request.active_conflicts or [],
        }
    ))

    if not result.success:
        raise HTTPException(status_code=500, detail=result.error)

    import json
    try:
        data = json.loads(result.output)
        return ConflictTrackingResponse(
            success=True,
            conflict_status=data.get("conflict_status", []),
            escalation_warnings=data.get("escalation_warnings", []),
            resolution_suggestions=data.get("resolution_suggestions", []),
        )
    except json.JSONDecodeError:
        raise HTTPException(status_code=500, detail="解析追踪结果失败")


@router.get("/conflicts/{conflict_id}")
async def get_conflict(conflict_id: str, project_id: str):
    """
    获取冲突线详情
    """
    # TODO: 从数据库获取
    raise HTTPException(status_code=404, detail="冲突线不存在")


@router.post("/conflicts/{conflict_id}/resolve")
async def resolve_conflict(conflict_id: str, chapter_number: int, resolution: str):
    """
    解决冲突线

    标记冲突为已解决
    """
    return {
        "success": True,
        "conflict_id": conflict_id,
        "status": "resolved",
        "resolved_at": datetime.now().isoformat(),
        "resolution_chapter": chapter_number,
        "resolution_description": resolution,
    }

File: frontend/src/api/villains.ts
/**
 * 反派管理 API 客户端
 * GodView v9: 反派与冲突系统
 */

import { api as apiClient } from './client';

// ==================== 类型定义 ====================

export type VillainLevel = 'small' | 'medium' | 'major';
export type ConflictStatus = 'active' | 'escalated' | 'resolved';
export type ConflictType = 'main' | 'sub' | 'hidden';

export interface Villain {
  id: string;
  project_id: string;
  name: string;
  alias: string | null;
  level: VillainLevel;
  description: string;
  motivation: string;
  goals: string[];
  strengths: string[];
  weaknesses: string[];
  relationship_with_protagonist: string;
  first_appearance_chapter: number | null;
  defeat_chapter: number | null;
  defeat_method: string | null;
  status: 'active' | 'defeated' | 'escaped' | 'redeemed';
  subordinates: string[];
  resources: string[];
  created_at: string;
}

export interface Conflict {
  id: string;
  project_id: string;
  conflict_type: ConflictType;
  title: string;
  description: string;
  parties: string[];
  stakes: string;
  status: ConflictStatus;
  intensity: number;
  start_chapter: number | null;
  resolution_chapter: number | null;
  resolution_method: string | null;
  escalation_history: ConflictEscalation[];
  created_at: string;
}

export interface ConflictEscalation {
  chapter: number;
  from_intensity: number;
  to_intensity: number;
  reason: string;
}

export interface VillainDesignRequest {
  project_id: string;
  level: VillainLevel;
  name?: string;
  genre?: string;
  protagonist_traits?: string[];
}

export interface CreateVillainRequest {
  project_id: string;
  name: string;
  level: VillainLevel;
  description: string;
  motivation: string;
  goals?: string[];
  strengths?: string[];
  weaknesses?: string[];
}

export interface CreateConflictRequest {
  project_id: string;
  conflict_type: ConflictType;
  title: string;
  description: string;
  parties: string[];
  stakes: string;
  start_chapter?: number;
}

// ==================== API 函数 ====================

/**
 * 获取反派列表
 */
export async function getVillains(projectId: string, level?: VillainLevel): Promise<Villain[]> {
  const params = new URLSearchParams({ project_id: projectId });
  if (level) params.append('level', level);
  const response = await apiClient.get(`/villains?${params.toString()}`);
  return response.data;
}

/**
 * AI 设计反派
 */
export async function designVillain(data: VillainDesignRequest): Promise<Villain> {
  const response = await apiClient.post('/villains/design', data);
  return response.data;
}

/**
 * 创建反派
 */
export async function createVillain(data: CreateVillainRequest): Promise<Villain> {
  const response = await apiClient.post('/villains', data);
  return response.data;
}

/**
 * 获取反派详情
 */
export async function getVillain(villainId: string): Promise<Villain> {
  const response = await apiClient.get(`/villains/${villainId}`);
  return response.data;
}

/**
 * 更新反派
 */
export async function updateVillain(villainId: string, data: Partial<CreateVillainRequest>): Promise<Villain> {
  const response = await apiClient.put(`/villains/${villainId}`, data);
  return response.data;
}

/**
 * 标记反派被击败
 */
export async function defeatVillain(villainId: string, chapterNumber: number, method: string): Promise<Villain> {
  const response = await apiClient.post(`/villains/${villainId}/defeat`, {
    chapter: chapterNumber,
    method,
  });
  return response.data;
}

/**
 * 获取冲突线列表
 */
export async function getConflicts(projectId: string, status?: ConflictStatus): Promise<Conflict[]> {
  const params = new URLSearchParams({ project_id: projectId });
  if (status) params.append('status', status);
  const response = await apiClient.get(`/villains/conflicts?${params.toString()}`);
  return response.data;
}

/**
 * 创建冲突线
 */
export async function createConflict(data: CreateConflictRequest): Promise<Conflict> {
  const response = await apiClient.post('/villains/conflicts', data);
  return response.data;
}

/**
 * 获取冲突线详情
 */
export async function getConflict(conflictId: string): Promise<Conflict> {
  const response = await apiClient.get(`/villains/conflicts/${conflictId}`);
  return response.data;
}

/**
 * 升级冲突
 */
export async function escalateConflict(conflictId: string, chapter: number, reason: string): Promise<Conflict> {
  const response = await apiClient.post(`/villains/conflicts/${conflictId}/escalate`, {
    chapter,
    reason,
  });
  return response.data;
}

/**
 * 解决冲突
 */
export async function resolveConflict(conflictId: string, chapter: number, method: string): Promise<Conflict> {
  const response = await apiClient.post(`/villains/conflicts/${conflictId}/resolve`, {
    chapter,
    method,
  });
  return response.data;
}

/**
 * 冲突追踪分析
 */
export async function trackConflicts(projectId: string, chapter?: number): Promise<{
  active_conflicts: Conflict[];
  escalation_warnings: string[];
  resolution_suggestions: string[];
}> {
  const params = new URLSearchParams({ project_id: projectId });
  if (chapter) params.append('chapter', String(chapter));
  const response = await apiClient.post(`/villains/tracking?${params.toString()}`);
  return response.data;
}

export default {
  getVillains,
  designVillain,
  createVillain,
  getVillain,
  updateVillain,
  defeatVillain,
  getConflicts,
  createConflict,
  getConflict,
  escalateConflict,
  resolveConflict,
  trackConflicts,
};
